File: test/frame-assembler.test.js
import { test } from "node:test";
import assert from "node:assert/strict";
import { createAssemblyState, advanceAssembly } from "../js/frame-assembler.js";
import { encodeMessage } from "../js/protocol.js";

function eventsForFrame(frame) {
  return Array.from(frame).map((index, i) => ({ index, afterLongGap: i === 0 }));
}

test("assembles a full valid frame into the decoded text", () => {
  const frame = encodeMessage("Hi");
  let state = createAssemblyState();
  let done = null;
  for (const event of eventsForFrame(frame)) {
    ({ state, done } = advanceAssembly(state, event));
  }
  assert.deepEqual(done, { ok: true, text: "Hi" });
  assert.deepEqual(state, createAssemblyState());
});

test("ignores symbols observed before the real transmission start", () => {
  let state = createAssemblyState();
  // ruido de fondo: simbolos sin pausa larga previa, no deberian arrancar nada
  ({ state } = advanceAssembly(state, { index: 55, afterLongGap: false }));
  ({ state } = advanceAssembly(state, { index: 3, afterLongGap: false }));
  assert.deepEqual(state, createAssemblyState());
});

test("starts accumulating once the long-gap start marker arrives", () => {
  let state = createAssemblyState();
  const result = advanceAssembly(state, { index: 2, afterLongGap: true });
  assert.deepEqual(result.state, { receiving: true, buffer: [2] });
  assert.equal(result.done, null);
});

test("reports a checksum-mismatch error for a corrupted frame and resets state", () => {
  const frame = encodeMessage("Hi");
  const corrupted = Uint8Array.from(frame);
  corrupted[1] ^= 0xff; // corrompe el primer byte del payload
  let state = createAssemblyState();
  let done = null;
  for (const event of eventsForFrame(corrupted)) {
    ({ state, done } = advanceAssembly(state, event));
  }
  assert.deepEqual(done, { ok: false, error: "checksum-mismatch" });
  assert.deepEqual(state, createAssemblyState());
});

test("can assemble a second message right after the first completes", () => {
  let state = createAssemblyState();
  for (const event of eventsForFrame(encodeMessage("Hi"))) {
    ({ state } = advanceAssembly(state, event));
  }
  let done = null;
  for (const event of eventsForFrame(encodeMessage("Chau"))) {
    ({ state, done } = advanceAssembly(state, event));
  }
  assert.deepEqual(done, { ok: true, text: "Chau" });
});

File: android/app/src/test/java/com/memetransfer/receiver/protocol/FrameAssemblerTest.kt
package com.memetransfer.receiver.protocol

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

private fun encodeMessageForTest(text: String): List<Int> {
    val payload = text.toByteArray(Charsets.UTF_8).map { it.toInt() and 0xFF }
    return payload + crc8(payload)
}

private fun eventsForFrame(frame: List<Int>): List<Any> =
    listOf(START) + frame + listOf(END)

class FrameAssemblerTest {
    @Test
    fun `assembles a full valid frame into the decoded text`() {
        var state = createAssemblyState()
        var done: DecodeResult? = null
        for (event in eventsForFrame(encodeMessageForTest("Hi"))) {
            val (nextState, result) = advanceAssembly(state, event)
            state = nextState
            if (result != null) done = result
        }
        assertEquals(DecodeResult.Ok("Hi"), done)
        assertEquals(createAssemblyState(), state)
    }

    @Test
    fun `ignores data bytes observed before the START marker`() {
        var state = createAssemblyState()
        state = advanceAssembly(state, 55).first
        state = advanceAssembly(state, 3).first
        assertEquals(createAssemblyState(), state)
    }

    @Test
    fun `ignores an END marker observed before any START`() {
        val state = createAssemblyState()
        val (next, done) = advanceAssembly(state, END)
        assertEquals(createAssemblyState(), next)
        assertNull(done)
    }

    @Test
    fun `START resets accumulation even mid-transmission (a resend always wins)`() {
        var state = createAssemblyState()
        state = advanceAssembly(state, START).first
        state = advanceAssembly(state, 1).first
        state = advanceAssembly(state, 2).first
        state = advanceAssembly(state, START).first // arranca de nuevo
        assertEquals(AssemblyState(receiving = true, buffer = emptyList()), state)
    }

    @Test
    fun `reports a checksum-mismatch error for a corrupted frame and resets state`() {
        val frame = encodeMessageForTest("Hi").toMutableList()
        frame[0] = frame[0] xor 0xff
        var state = createAssemblyState()
        var done: DecodeResult? = null
        for (event in eventsForFrame(frame)) {
            val (nextState, result) = advanceAssembly(state, event)
            state = nextState
            if (result != null) done = result
        }
        assertTrue(done is DecodeResult.Error && (done as DecodeResult.Error).error == "checksum-mismatch")
        assertEquals(createAssemblyState(), state)
    }

    @Test
    fun `can assemble a second message right after the first completes`() {
        var state = createAssemblyState()
        for (event in eventsForFrame(encodeMessageForTest("Hi"))) {
            state = advanceAssembly(state, event).first
        }
        var done: DecodeResult? = null
        for (event in eventsForFrame(encodeMessageForTest("Chau"))) {
            val (nextState, result) = advanceAssembly(state, event)
            state = nextState
            if (result != null) done = result
        }
        assertEquals(DecodeResult.Ok("Chau"), done)
    }
}
